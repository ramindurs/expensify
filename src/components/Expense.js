import React from 'react';
import {Link} from 'react-router-dom';
import moment from "moment";
import numeral from 'numeral';

export const Expense = ({ id, description, amount, createdAt, dispatch, history}) => {
    return (
        <div>
            {id} <Link to={`/edit/${id}` }>{description}</Link>
            <p>
                {numeral(amount / 100).format('0,0.00')}
                -
                {moment(createdAt).format('MMM Do YYYY')}
            </p>
        </div>
    );
};

export default Expense;
