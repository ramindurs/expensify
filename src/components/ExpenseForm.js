import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from "react-dates";


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        const expense = props.expenseToEdit;
        this.state = {
            description: expense && expense.description ? expense.description : '',
            note: expense && expense.note ? expense.note : '',
            amount: expense && expense.amount ? this.convertIntToFloatString(expense.amount) : '',
            createdAt: expense && expense.createdAt ? moment(expense.createdAt) : moment(),
            calendarFocus: false,
            error: ''
        };
    }
    convertIntToFloatString = (intnum) => parseFloat(intnum) / 100;

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    };
    onDateChange = (date) => {
        if (!!date) {
            this.setState(() => ({createdAt: date}));
        }
    };
    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocus: focused}));
    };
    onSubmitForm = (e) => {
        e.preventDefault();
        if (this.state.description && this.state.amount) {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                amount: Math.round(parseFloat(this.state.amount, 10) * 100),
                createdAt: this.state.createdAt.valueOf()
            });
        } else {
            this.setState(() => ({error: 'Please provide description and amount'}));
        }
    };

    render() {
        return (
            <div>
                {this.state.error ? <p>{this.state.error}</p> : ''}
                <form onSubmit={this.onSubmitForm}>
                    <input
                        type="text"
                        placeholder={'description'}
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        autoFocus/>
                    <input
                        type="text"
                        placeholder={'amount'}
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocus}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        onChange={this.onNoteChange}
                        value={this.state.note}
                        placeholder={'Add a note'}>
                    </textarea>
                    <button>{this.props.submitMessage}</button>
                </form>
            </div>
        )
    };
};
