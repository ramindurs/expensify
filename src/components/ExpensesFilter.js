import React from 'react';
import {connect} from 'react-redux';
import { DateRangePicker} from "react-dates";
import {setSortByAmount, setSortByDate, setTextFilter, setStartDate, setEndDate} from "../actions/filters";

export class ExpensesFilter extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onTextFilterChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    setSortBy = (e) => {
        const sortBy = e.target.value;
        switch (sortBy) {
            case 'amount':
                this.props.setSortByAmount();
                return;
            default:
                this.props.setSortByDate();
                return;
        }
    };

    onFocusChange = (focusedInput) => {
        this.setState(() => ({ calendarFocused: focusedInput}));
    };

    render() {
        return (
            <div className='content_container'>
                <div className="input-group">
                    <div className='input-group__item'>
                        <input
                            className="text-input"
                            placeholder='Search expenses'
                            type="text"
                            value={this.props.filters.text}
                            onChange={this.onTextFilterChange}/>
                    </div>
                    <div>
                        <select
                            className='select'
                            value={this.props.filters.sortBy}
                            onChange={this.setSortBy}>
                            <option value={'date'}>Date</option>
                            <option value={'amount'}>Amount</option>
                        </select>
                    </div>
                    <div>
                        <DateRangePicker
                            startDateId={'startDate'}
                            startDate={this.props.filters.startDate}
                            endDateId={'endDate'}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            showClearDates={true}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

const mappedStateToProps = (state) => {
    return {
        filters: state.filters
    }
};

const mappedDispatchToProps = (dispatch) => {
    return {
        setStartDate: (date) => dispatch(setStartDate(date)),
        setEndDate: (date) => dispatch(setEndDate(date)),
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        setSortByAmount: () => dispatch(setSortByAmount()),
        setSortByDate: () => dispatch(setSortByDate())
    };
};

export default connect(mappedStateToProps, mappedDispatchToProps)(ExpensesFilter);