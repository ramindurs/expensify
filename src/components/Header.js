import React from 'react';
import {Link} from "react-router-dom";
import {connect} from 'react-redux';
import {startLogout} from "../actions/auth";

export const Header = ({startLogout}) => (
    <header className="header-layout">
        <div className="content_container">
            <div className="header-layout__content">
                <Link className="header-layout__title" to={"/dashboard"}>
                    <h1>Expensify</h1>
                </Link>
                <button className="box-layout--link" onClick={startLogout}>Logout</button>
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
        startLogout: () => (dispatch(startLogout()))
    }
);

export default connect(undefined, mapDispatchToProps)(Header);
