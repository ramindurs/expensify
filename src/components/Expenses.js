import React from 'react';
import {connect} from "react-redux";
import Expense from "./Expense";
import selectExpenses from '../selectors/expenses';

export const Expenses = (props) => (
    <div className="content_container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ?
                    (<div className="list-item list-item--message">
                        <span>No expenses</span>
                    </div>) :
                    (props.expenses.map((expense, index) =>
                        <Expense key={index}
                                 {...expense}
                        />))
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(Expenses);
