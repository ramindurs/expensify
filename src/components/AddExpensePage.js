import React from 'react';
import {connect} from 'react-redux';
import {startAddExpense} from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                <div className='page-header-layout'>
                    <div className="content_container">
                        <h1 className='page-header-layout__title'>Add Expenses</h1>
                    </div>
                </div>
                <div className="content_container">
                    <ExpenseForm
                        submitMessage={'Save Expense'}
                        onSubmit={this.onSubmit}/>
                </div>
            </div>
        );
    }
}

const mapDispatchTpProps = (dispatch) => {
    return {
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
    };
};

export default connect(undefined, mapDispatchTpProps)(AddExpensePage);
