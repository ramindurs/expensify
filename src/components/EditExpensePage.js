import React from 'react';
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {startEditExpense, startRemoveExpense} from "../actions/expenses";

export class EditExpensePage extends React.Component {
    removeExpense = (id) => {
        this.props.startRemoveExpense(id);
        this.props.history.push('/dashboard');
    };
    editExpense = (id, expense) => {
        this.props.startEditExpense(id, expense);
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                <div className="page-header-layout">
                    <div className="content_container">
                        <h1 className="page-header-layout__title">
                            Editing Expense
                            <span className="page-header-layout__title-reduce">
                                &nbsp; &nbsp; ID {this.props.expense.id}
                            </span>
                        </h1>
                    </div>
                </div>
                <div className="content_container">
                    <ExpenseForm
                        expenseToEdit={this.props.expense}
                        submitMessage={'Save Updates'}
                        onSubmit={(expense) => {
                            this.editExpense(this.props.expense.id, expense)
                        }}
                    />
                    <button
                        className="button-layout__danger"
                        onClick={() => {
                            this.removeExpense(this.props.expense.id)
                        }}>
                        Delete Expense
                    </button>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(e => e.id === props.match.params.id)
    };
};

const matchDispatchToProps = (dispatch, props) => {
    return {
        startEditExpense: (id, expense) => dispatch(startEditExpense({
            id: id,
            updates: expense
        })),
        startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
    };
};

export default connect(mapStateToProps, matchDispatchToProps)(EditExpensePage);
