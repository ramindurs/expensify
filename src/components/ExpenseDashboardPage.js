import React from 'react';
import Expenses from "./Expenses";
import ExpensesFilter from "./ExpensesFilter";
import ExpensesSummary from "./ExpensesSummary";

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary/>
        <ExpensesFilter/>
        <Expenses/>
    </div>
);

export default ExpenseDashboardPage;