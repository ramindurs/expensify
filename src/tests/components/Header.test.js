import React from 'react';
import {shallow} from 'enzyme';
import {Header} from "../../components/Header";

test('should render header correctly', () => {
    const wrapper = shallow(<Header startLogout={() => {
    }}/>);

    expect(wrapper).toMatchSnapshot();
});

test('should call startLogout on logout button clicked', () => {
    const logoutSpy = jest.fn();
    const wrapper = shallow(<Header startLogout={logoutSpy}/>);

    wrapper.find('button').simulate('click');

    expect(logoutSpy).toHaveBeenCalled();
});

