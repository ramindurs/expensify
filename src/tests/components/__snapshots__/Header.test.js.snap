// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header correctly 1`] = `
<header
  className="header-layout"
>
  <div
    className="content_container"
  >
    <div
      className="header-layout__content"
    >
      <Link
        className="header-layout__title"
        to="/dashboard"
      >
        <h1>
          Expensify
        </h1>
      </Link>
      <button
        className="box-layout--link"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
